crates/resolver/src/hosts.rs

```
@@ -109,69 +109,73 @@ fn hosts_path() -> std::path::PathBuf {
 #[cfg_attr(docsrs, doc(cfg(any(unix, windows))))]
 pub(crate) fn read_hosts_conf<P: AsRef<Path>>(path: P) -> io::Result<Hosts> {
     use std::fs::File;
-    use std::io::{BufRead, BufReader};
 
-    use proto::rr::domain::TryParseIp;
-
-    let mut hosts = Hosts {
-        by_name: HashMap::new(),
-    };
-
-    // lines in the file should have the form `addr host1 host2 host3 ...`
-    // line starts with `#` will be regarded with comments and ignored,
-    // also empty line also will be ignored,
-    // if line only include `addr` without `host` will be ignored,
-    // file will parsed to map in the form `Name -> LookUp`.
     let file = File::open(path)?;
+    Ok(Hosts::default().read_hosts_conf(file))
+}
 
-    for line in BufReader::new(file).lines() {
-        // Remove comments from the line
-        let line = line
-            .as_ref()
-            .map(|line| line.split('#').next().unwrap().trim())
-            .unwrap_or_default();
-        if line.is_empty() {
-            continue;
-        }
+impl Hosts {
+    /// parse configuration from `src`
+    pub fn read_hosts_conf(mut self, src: impl io::Read) -> Self {
+        use std::io::{BufRead, BufReader};
+
+        use proto::rr::domain::TryParseIp;
+
+        // lines in the src should have the form `addr host1 host2 host3 ...`
+        // line starts with `#` will be regarded with comments and ignored,
+        // also empty line also will be ignored,
+        // if line only include `addr` without `host` will be ignored,
+        // the src will be parsed to map in the form `Name -> LookUp`.
+
+        for line in BufReader::new(src).lines() {
+            // Remove comments from the line
+            let line = line
+                .as_ref()
+                .map(|line| line.split('#').next().unwrap().trim())
+                .unwrap_or_default();
```

Let's return this error.